### DISCUSSION TITLE
[Feature request] [3.0] Add support for custom renderers

### DISCUSSION THREAD

**Discussion by ItsCubeTime** [UPBGE/upbge#1712] [Ideas/Features for UPBGE 0.3+]
Hi everyone!

https://github.com/dfelinto/blender/blob/master/doc/python_api/examples/bpy.types.RenderEngine.py is a  python script that registers a custom render engine.

Currently what happens if you switch over to this template render engine and attempt to start an embedded or standalone game, it switches to Eevee in the newly launched window. Starting the game with embedded renderer appears to work like expected on the other hand.

Another issue is that the Game Resolution & Game Debug panels wont automatically register to custom render engines:
![image](https://user-images.githubusercontent.com/20190653/159118400-74d49b63-d4ef-4cfa-9108-74d8a13ea97b.png)

A good motivation for adding support for custom renderers would be for individual projects to be able to use renderers that fit there artstyle and game specific needs better. Eevee as it stands is not very optimized for games, with its rasterizer often taking up 90% or more of the total frame time making it difficult to use for relatively complex scenes. The use of custom renderers would unlock users to use (by use I mean port) existing more optimized renderers or write there own and I believe that would unlock a lot more potential of the game engine in general and in the long term hopefully also some community driven renderers that can make a good alternative to Eevee for the average UPBGE user (I could see myself take on a project like this in the near future, which is why Im making this request now 😄 ).

Ideally standalone debug sessions or exported builds should use the currently active render engine by default without any hackarounds

Thanks for reading!

**Comment by youle31** [comment]
- When you don't use "viewport render" mode, the custom bge render loop is written to use eevee (either lookdev or eevee render mode) because eevee is the render engine which looks like the more appropriate to do small games even if it has its limits.
If game is started from solid mode (workbench), it will automatically switch to eevee rendered mode.
If game is started from lookdev, it will stay in lookdev.
If game is started from eevee rendered, it will stay in eevee rendered mode.

It is because when we start standalone or exported game:

1. If .blend is saved in solid mode (which is not really appropriate for game), we want the game to be opened with a more appropriate engine (eevee).
2. If .blend is saved in lookdev, we want the game to be opened in lookdev which can be used to do small games.
3. .blend saved while we are in eevee rendered mode are normally opened in solid view (this is the way blender works), and as solid view is not really appropriate for game, we force the switch to eevee rendered mode.

- When you use "viewport render" mode, there are similar reasons which explain the choices done to "force" a render engine when you start launching "embedded player" or "standalone player", but you can normally bypass this with a simple bpy script to switch of engine once the game has been started: https://blender.stackexchange.com/questions/5291/how-to-switch-rendering-engine-via-python-script
It might not work properly with standalone player in all cases (because not tested) but it should work at least in embedded I think (you can use cycles in embedded for example, and I guess you can use other render engines too with a bpy script launched after game start).

**Comment by lordloki** [comment]
Moving to discussions
